packages/devtools/lib/src/timeline/frame_flame_chart.dart

```diff
@@ -407,3 +430,142 @@ class FlameChartItem {
         colorToCss(selected ? selectedTextColor : defaultTextColor);
   }
 }
+
+class TimelineGrid extends CoreElement {
+  TimelineGrid(this._frameDurationMs, this.flameChartWidth)
+      : super('div', classes: 'flame-chart-grid') {
+    flex();
+    layoutHorizontal();
+    _drawGrid(baseGridInterval);
+  }
+
+  static const baseGridInterval = 150;
+
+  final num _frameDurationMs;
+
+  num zoomLevel = 1;
+
+  num flameChartWidth;
+
+  num get flameChartWidthWithInsets => flameChartWidth + 2 * flameChartInset;
+
+  num interval = baseGridInterval;
+
+  final List<TimelineGridItem> _gridItems = [];
+
+  void _drawGrid(num interval) {
+    // Draw the first grid item since it will have a different width than the
+    // rest.
+    final gridItem = TimelineGridItem(0, flameChartInset, 0);
+    _gridItems.add(gridItem);
+    add(gridItem);
+
+    num left = flameChartInset;
+
+    while (left + interval < flameChartWidthWithInsets) {
+      final timestamp = getTimestampForPosition(left + interval);
+      final gridItem = TimelineGridItem(left, interval, timestamp);
+
+      _gridItems.add(gridItem);
+      add(gridItem);
+
+      left += interval;
+    }
+  }
+
+  void _clear() {
+    clear();
+    _gridItems.clear();
+  }
+
+  num getTimestampForPosition(num gridItemEnd) {
+    return (gridItemEnd - flameChartInset) / flameChartWidth * _frameDurationMs;
+  }
+
+  void updateForZoom(num newZoomLevel, num newFlameChartWidth) {
+    if (zoomLevel == newZoomLevel) {
+      return;
+    }
+
+    flameChartWidth = newFlameChartWidth;
+    element.style.width = '${flameChartWidthWithInsets}px';
+
+    final log2ZoomLevel = log2(zoomLevel);
+    final log2NewZoomLevel = log2(newZoomLevel);
+
+    final gridZoomFactor = math.pow(2, log2NewZoomLevel);
+    final gridIntervalPx = baseGridInterval / gridZoomFactor;
+
+    /// The physical pixel width of the grid interval at [newZoomLevel].
+    final zoomedGridIntervalPx = gridIntervalPx * newZoomLevel;
+
+    if (log2NewZoomLevel == log2ZoomLevel) {
+      // Don't modify the first grid item. This item will have a fixed left of
+      // 0, width of [flameChartInset], and timestamp of '0.0 ms'.
+      for (int i = 1; i < _gridItems.length; i++) {
+        final currentItem = _gridItems[i];
+
+        final newLeft = flameChartInset + zoomedGridIntervalPx * (i - 1);
+        currentItem.update(newLeft, zoomedGridIntervalPx);
+      }
+    } else {
+      _clear();
+      _drawGrid(zoomedGridIntervalPx);
```

Comment:
add a TODO to test this grid drawing logic.

Reply:
done.